lightning-block-sync/src/init.rs

```diff
@@ -0,0 +1,427 @@
+/// 		let read_args = ChannelManagerReadArgs::new(
+/// 			keys_manager,
+/// 			fee_estimator,
+/// 			chain_monitor,
+/// 			tx_broadcaster,
+/// 			logger,
+/// 			config,
+/// 			vec![&mut monitor],
+/// 		);
+/// 		<(BlockHash, ChannelManager<S, &ChainMonitor<S, &C, &T, &F, &L, &P>, &T, &K, &F, &L>)>::read(
+/// 			&mut Cursor::new(&serialized_manager), read_args).unwrap()
+/// 	};
+///
+/// 	let mut cache = UnboundedCache::new();
+/// 	let mut monitor_listener = (RefCell::new(monitor), tx_broadcaster, fee_estimator, logger);
+/// 	let mut manager_listener = &manager;
+/// 	let listeners = vec![
+/// 		(monitor_block_hash, &mut monitor_listener as &mut dyn ChainListener),
+/// 		(manager_block_hash, &mut manager_listener as &mut dyn ChainListener),
+/// 	];
+/// 	let chain_tip =
+/// 		init::sync_listeners(block_source, Network::Bitcoin, &mut cache, listeners).await.unwrap();
+///
+/// 	let monitor = monitor_listener.0.into_inner();
+/// 	chain_monitor.watch_channel(monitor.get_funding_txo().0, monitor);
+///
+/// 	let chain_poller = poll::ChainPoller::new(block_source, Network::Bitcoin);
+/// 	let chain_listener = (&chain_monitor, &manager_listener);
+/// 	let spv_client = SpvClient::new(chain_tip, chain_poller, &mut cache, chain_listener);
+/// }
+/// ```
+///
+/// [`SpvClient`]: ../struct.SpvClient.html
+/// [`ChannelManager`]: ../../lightning/ln/channelmanager/struct.ChannelManager.html
+/// [`ChannelMonitor`]: ../../lightning/chain/channelmonitor/struct.ChannelMonitor.html
+pub async fn sync_listeners<B: BlockSource, C: Cache>(
+	block_source: &mut B,
+	network: Network,
+	header_cache: &mut C,
+	mut chain_listeners: Vec<(BlockHash, &mut dyn ChainListener)>,
+) -> BlockSourceResult<ValidatedBlockHeader> {
+	let (best_block_hash, best_block_height) = block_source.get_best_block().await?;
+	let new_header = block_source
+		.get_header(&best_block_hash, best_block_height).await?
+		.validate(best_block_hash)?;
+
+	// Fetch the header for the block hash paired with each listener.
+	let mut chain_listeners_with_old_headers = Vec::new();
+	for (old_block, chain_listener) in chain_listeners.drain(..) {
+		let old_header = match header_cache.look_up(&old_block) {
+			Some(header) => *header,
+			None => block_source
+				.get_header(&old_block, None).await?
+				.validate(old_block)?
+		};
+		chain_listeners_with_old_headers.push((old_header, chain_listener))
+	}
+
+	// Find differences and disconnect blocks for each listener individually.
+	let mut chain_poller = ChainPoller::new(block_source, network);
+	let mut chain_listeners_at_height = Vec::new();
+	let mut most_common_ancestor = None;
+	let mut most_connected_blocks = Vec::new();
+	for (old_header, chain_listener) in chain_listeners_with_old_headers.drain(..) {
+		// Disconnect any stale blocks, but keep them in the cache for the next iteration.
+		let header_cache = &mut ReadOnlyCache(header_cache);
+		let mut chain_notifier = ChainNotifier { header_cache };
+		let difference =
```

Comment:
IIUC `find_difference` if the old block header isn't on same chain, we'll try to connect blocks anyway leading to an invalid, mixed chain of block connections from the viewpoint of the listener ?

I think this case is catched-up by `get_header` on block source above, but I think it should be documented in top-down documentation, what a user is expected to do if you hit that kind of failure ?

Reply:
are you saying if the new chain and the stale chain have no common ancestor? Is that not a rather ridiculously unlikely scenario?

Reply:
`find_difference` will fail in this case as covered by the `sync_from_different_chains` unit test.

Reply:
@arik-so a belt-and-suspender if someone screws up between regtest/testnet/mainet ? But like I say already covered, good to me